README.rst
====================================================
Community Detection Enrichr Gene to Term Mapper
====================================================

This repository creates a CDAPS compatible functional enrichment Docker image using Enrichr
packaged from https://pypi.org/project/gseapy/

Dependencies
------------

* `Docker <https://www.docker.com/>`_
* `make <https://www.gnu.org/software/make/>`_ (to build)
* Python (to build)

Direct invocation
------------------

Version `0.4.0` can be directly pulled from `Dockerhub <https://hub.docker.com/>`_ with this command:

.. code-block::

   docker pull coleslawndex/cdenrichrgenestoterm:0.4.0

Building
--------

.. code-block::

   git clone https://github.com/idekerlab/cdenrichrgenestoterm
   cd cdenrichrgenestoterm
   make dockerbuild

Run **make** command with no arguments to see other build/deploy options including creation of Docker image

.. code-block::

   make

Output:

.. code-block::

   clean                remove all build, test, coverage and Python artifacts
   clean-build          remove build artifacts
   clean-pyc            remove Python file artifacts
   clean-test           remove test and coverage artifacts
   lint                 check style with flake8
   test                 run tests quickly with the default Python
   test-all             run tests on every Python version with tox
   coverage             check code coverage quickly with the default Python
   docs                 generate Sphinx HTML documentation, including API docs
   servedocs            compile the docs watching for changes
   testrelease          package and upload a TEST release
   release              package and upload a release
   dist                 builds source and wheel package
   install              install the package to the active Python's site-packages
   dockerbuild          build docker image and store in local repository
   dockerpush           push image to dockerhub


Usage
-----

.. code-block::

   docker run -v coleslawndex/cdenrichrgenestoterm:0.4.0 -h



Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.4.0 (2021-03-09)
----------------------

* Updated to use gseapy 0.10.4 because Enrichr API changed hostname
  and older versions no longer worked 
  `Issue #2 <https://github.com/idekerlab/cdenrichrgenestoterm/issues/2>`_

0.3.1 (2020-08-13)
----------------------

* Removed debugging print statement that was sending debugging output
to standard output causing CDAPS app to fail.

* Fixed bug where best hit from last gene set input was
  returned instead of best hit from any of the gene sets

0.2.0 (2019-12-13)
--------------------

* Added term_size which denotes number of genes associated with term to output JSON

0.1.0 (2019-09-19)
------------------

* First release on PyPI.


tests/test_cdenrichrgenestoterm.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_cdenrichrgenestoterm
----------------------------------

Tests for `cdenrichrgenestoterm` module.
"""

import os
import sys
import unittest
import tempfile
import shutil
from unittest.mock import MagicMock, call
import pandas as pd


from cdenrichrgenestoterm import cdenrichrgenestoterm


class TestCdenrichrgenestoterm(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def get_default_genesets(self):
        return 'GO_Biological_Process_2018,' \
               'GO_Cellular_Component_2018,' \
               'GO_Molecular_Function_2018'

    def test_read_inputfile(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('hellothere')
            res = cdenrichrgenestoterm.read_inputfile(tfile)
            self.assertEqual('hellothere', res)
        finally:
            shutil.rmtree(temp_dir)

    def test_parse_args(self):
        myargs = ['inputarg']
        res = cdenrichrgenestoterm._parse_arguments('desc',
                                                      myargs)
        self.assertEqual('inputarg', res.input)
        self.assertEqual(0.05, res.maxpval)
        self.assertEqual('/tmp', res.tmpdir)
        self.assertEqual('GO_Biological_Process_2018,' +
                         'GO_Cellular_Component_2018,' +
                         'GO_Molecular_Function_2018',
                         res.genesets)

    def test_run_gprofiler_no_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = [tfile]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                              myargs)
            try:
                cdenrichrgenestoterm.run_enrichr(tfile,
                                                 theargs)
                self.fail('Expected FileNotFoundError')
            except FileNotFoundError:
                pass
        finally:
            shutil.rmtree(temp_dir)

    def test_run_gprofiler_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = [tfile]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                              myargs)
            res = cdenrichrgenestoterm.run_enrichr(tfile,
                                                   theargs)
            self.assertEqual(None, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_with_empty_result(self):
        temp_dir = tempfile.mkdtemp()
        try:
            enrichr = MagicMock()
            enrichr.enrichr = MagicMock(return_value=enrichr)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c')
            myargs = [tfile, '--tmpdir', temp_dir]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                            myargs)
            res = cdenrichrgenestoterm.run_enrichr(tfile,
                                                   theargs,
                                                   enrichr=enrichr)
            gs = self.get_default_genesets()
            self.assertEqual(None, res)
            enrichr.enrichr.assert_called_once_with(gene_list=['A', 'B', 'C'],
                                                    cutoff=0.05,
                                                    gene_sets=gs,
                                                    no_plot=True,
                                                    outdir=temp_dir)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_where_enrichr_raises_exception(self):
        temp_dir = tempfile.mkdtemp()
        try:
            enrichr = MagicMock()
            enrichr.enrichr = MagicMock()
            enrichr.enrichr.side_effect = Exception('Some exception')
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c')
            myargs = [tfile, '--tmpdir', temp_dir]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                            myargs)
            res = cdenrichrgenestoterm.run_enrichr(tfile,
                                                   theargs,
                                                   enrichr=enrichr)
            gs = self.get_default_genesets()
            self.assertEqual(None, res)
            the_call = call(gene_list=['A', 'B', 'C'],
                            cutoff=0.05,
                            gene_sets=gs,
                            no_plot=True,
                            outdir=temp_dir)

            enrichr.enrichr.assert_has_calls([the_call, the_call])
        finally:
            shutil.rmtree(temp_dir)

    def test_run_with_valid_no_result_from_query(self):
        temp_dir = tempfile.mkdtemp()
        try:
            enrichr = MagicMock()
            enrichr.enrichr = MagicMock(return_value=enrichr)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c')
            myargs = [tfile, '--tmpdir', temp_dir]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                            myargs)
            res = cdenrichrgenestoterm.run_enrichr(tfile,
                                                   theargs,
                                                   enrichr=enrichr)
            gs = self.get_default_genesets()
            self.assertEqual(None, res)
            enrichr.enrichr.assert_called_once_with(gene_list=['A', 'B', 'C'],
                                                    cutoff=0.05,
                                                    gene_sets=gs,
                                                    no_plot=True,
                                                    outdir=temp_dir)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_with_pvalue_exceeded(self):
        temp_dir = tempfile.mkdtemp()
        try:
            enrichr = MagicMock()

            df = pd.DataFrame(columns=['Term',
                                       'Gene_set',
                                       'P-value',
                                       'Adjusted P-value',
                                       'Genes', 'Overlap'],
                              data=[['term1', 'set1', 1.5,
                                     0.6,
                                     'B;C', '2/25'],
                                    ['term2', 'set2', 1.6,
                                     0.7, 'A;C', '7/9']])

            txt_file = os.path.join(temp_dir, 'data.txt')
            df.to_csv(txt_file, index=False, sep='\t', encoding='utf-8')

            enrichr.enrichr = MagicMock(return_value=enrichr)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c')
            myargs = [tfile, '--tmpdir', temp_dir]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                            myargs)
            res = cdenrichrgenestoterm.run_enrichr(tfile,
                                                   theargs,
                                                   enrichr=enrichr)
            gs = self.get_default_genesets()
            self.assertEqual(None, res)
            enrichr.enrichr.assert_called_once_with(gene_list=['A', 'B', 'C'],
                                                    cutoff=0.05,
                                                    gene_sets=gs,
                                                    no_plot=True,
                                                    outdir=temp_dir)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_with_valid_result(self):
        temp_dir = tempfile.mkdtemp()
        try:
            enrichr = MagicMock()
            df = pd.DataFrame(columns=['Term',
                                       'Gene_set',
                                       'P-value',
                                       'Adjusted P-value',
                                       'Genes', 'Overlap'],
                              data=[['term1', 'set1', 0.6,
                                     0.05,
                                     'B;C', '2/25'],
                                    ['term2', 'set2', 0.5,
                                     0.03, 'A;C', '7/9']])
            enrichr.enrichr = MagicMock(return_value=enrichr)
            txt_file = os.path.join(temp_dir, 'data.txt')
            df.to_csv(txt_file, index=False, sep='\t', encoding='utf-8')
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c')
            myargs = [tfile, '--tmpdir', temp_dir]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                            myargs)
            res = cdenrichrgenestoterm.run_enrichr(tfile,
                                                   theargs,
                                                   enrichr=enrichr)
            gs = self.get_default_genesets()
            self.assertEqual('term2', res['name'])
            self.assertEqual('set2', res['source'])
            self.assertEqual(0.03, res['p_value'])
            self.assertEqual('', res['description'])
            self.assertEqual(['A', 'C'], res['intersections'])
            self.assertEqual(9, res['term_size'])
            enrichr.enrichr.assert_called_once_with(gene_list=['A', 'B', 'C'],
                                                    cutoff=0.05,
                                                    gene_sets=gs,
                                                    no_plot=True,
                                                    outdir=temp_dir)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_with_valid_result_from_multiple_genesets(self):
        temp_dir = tempfile.mkdtemp()
        try:
            enrichr = MagicMock()
            df = pd.DataFrame(columns=['Term',
                                       'Gene_set',
                                       'P-value',
                                       'Adjusted P-value',
                                       'Genes', 'Overlap'],
                              data=[['term1', 'set1', 0.6,
                                     0.05,
                                     'B;C', '2/25'],
                                    ['term2', 'set2', 0.5,
                                     0.03, 'A;C', '7/9']])

            txt_file = os.path.join(temp_dir, 'data.txt')
            df.to_csv(txt_file, index=False, sep='\t', encoding='utf-8')

            df = pd.DataFrame(columns=['Term',
                                       'Gene_set',
                                       'P-value',
                                       'Adjusted P-value',
                                       'Genes', 'Overlap'],
                              data=[['term5', 'set1', 0.6,
                                     0.01,
                                     'B;C', '2/25'],
                                    ['term6', 'set2', 0.5,
                                     0.03, 'X;Y', '7/8']])

            txt_file = os.path.join(temp_dir, 'data2.txt')
            df.to_csv(txt_file, index=False, sep='\t', encoding='utf-8')

            # make a directory with .txt ending just for fun
            os.makedirs(os.path.join(temp_dir, 'haha.txt'), mode=0o755)

            enrichr.enrichr = MagicMock(return_value=enrichr)
            tfile = os.path.join(temp_dir, 'foo')
            with open(tfile, 'w') as f:
                f.write('a,b,c')
            myargs = [tfile, '--tmpdir', temp_dir]
            theargs = cdenrichrgenestoterm._parse_arguments('desc',
                                                            myargs)
            res = cdenrichrgenestoterm.run_enrichr(tfile,
                                                   theargs,
                                                   enrichr=enrichr)
            gs = self.get_default_genesets()
            self.assertEqual('term5', res['name'])
            self.assertEqual('set1', res['source'])
            self.assertEqual(0.01, res['p_value'])
            self.assertEqual('', res['description'])
            self.assertEqual(['B', 'C'], res['intersections'])
            self.assertEqual(25, res['term_size'])
            enrichr.enrichr.assert_called_once_with(gene_list=['A', 'B', 'C'],
                                                    cutoff=0.05,
                                                    gene_sets=gs,
                                                    no_plot=True,
                                                    outdir=temp_dir)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_invalid_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = ['prog', tfile]
            res = cdenrichrgenestoterm.main(myargs)
            self.assertEqual(2, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = ['prog', tfile]
            res = cdenrichrgenestoterm.main(myargs)
            self.assertEqual(0, res)
        finally:
            shutil.rmtree(temp_dir)


if __name__ == '__main__':
    sys.exit(unittest.main())


tests/test_integration_of_cdenrichr.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_integration_of_enrichr
----------------------------------

Integration tests for `enrichr` module.
"""

import os
import sys
import unittest
import tempfile
import shutil
import json
import subprocess


SKIP_REASON = 'CDENRICHR_DOCKER_IMAGE, CDENRICHR_DOCKER, CDENRICHR_TMPDIR environment ' \
              'variable(s) not set to a docker image' \
              ' cannot run integration tests of Enrichr with Docker'


@unittest.skipUnless(os.getenv('CDENRICHR_DOCKER_IMAGE') is not None and
                     os.getenv('CDENRICHR_DOCKER') is not None and
                     os.getenv('CDENRICHR_TMPDIR') is not None, SKIP_REASON)
class TestCdhidefInDocker(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def run_enrichr_docker(self, cmdargs, temp_dir=None):
        """
        Runs hidef command as a command line process
        :param cmd_to_run: command to run as list
        :type cmd_to_run: list
        :return: (return code, standard out, standard error)
        :rtype: tuple
        """
        cmd = [os.getenv('CDENRICHR_DOCKER'), 'run', '--rm', '-v',
               temp_dir+':'+temp_dir,
               os.getenv('CDENRICHR_DOCKER_IMAGE')]
        cmd.extend(cmdargs)
        p = subprocess.Popen(cmd,
                             stdout=subprocess.PIPE,
                             stderr=subprocess.PIPE)

        out, err = p.communicate()

        return p.returncode, out, err

    def test_run_enrichr_small_gene_list_may_take_one_minute(self):
        temp_dir = tempfile.mkdtemp(dir=os.getenv('CDENRICHR_TMPDIR'))
        try:
            gene_list = 'NR1H3,MNDA,APOE,NCF2,CRIP3,NCF1,NCF4,CD48,LILRA6,' \
                        'LILRA4,CD44,LILRA5,CD47,LILRA2,LILRA1,ADA2,CD40,' \
                        'LINGO3,RAB20,CD14,MPP1,CD19,PTPN7,PTPN6,FTHL17,' \
                        'LHFPL2,UNC13D,COL23A1'
            input_file = os.path.join(temp_dir, 'input.txt')
            with open(input_file, 'w') as f:
                f.write(gene_list + '\n')
            ecode, out, err = self.run_enrichr_docker([input_file],
                                                      temp_dir=temp_dir)
            self.assertEqual(0, ecode)
            res = json.loads(out)
            self.assertEqual(8, len(res.keys()))
            self.assertEqual('secondary lysosome (GO:0005767)', res['name'])
            self.assertEqual('GO_Cellular_Component_2018', res['source'])
            self.assertEqual('', res['description'])
            self.assertEqual(13, res['term_size'])
            self.assertEqual(3, len(res['intersections']))

        finally:
            shutil.rmtree(temp_dir)

    def test_run_enrichr_two_gene_items_may_take_one_minute(self):
        temp_dir = tempfile.mkdtemp(dir=os.getenv('CDENRICHR_TMPDIR'))
        try:
            gene_list = 'MTOR,TP53'
            input_file = os.path.join(temp_dir, 'input.txt')
            with open(input_file, 'w') as f:
                f.write(gene_list + '\n')
            ecode, out, err = self.run_enrichr_docker([input_file,
                                                       '--genesets',
                                                       'Jensen_DISEASES'],
                                                      temp_dir=temp_dir)
            self.assertEqual(0, ecode)

            res = json.loads(out)
            self.assertEqual(8, len(res.keys()))
            self.assertEqual('Neurofibromatosis', res['name'])
            self.assertEqual('Jensen_DISEASES', res['source'])
            self.assertEqual('', res['description'])
            self.assertEqual(38, res['term_size'])
            self.assertEqual(2, len(res['intersections']))

        finally:
            shutil.rmtree(temp_dir)

    def test_run_enrichr_ud1214_items_may_take_one_minute(self):
        temp_dir = tempfile.mkdtemp(dir=os.getenv('CDENRICHR_TMPDIR'))
        try:
            gene_list = 'NPPB,NPR3,NPPC,NPR1,CRP,ELF1,NPR2,PAX5,NPPA,MME,SMUG1,ELF2,GUCY2C'
            input_file = os.path.join(temp_dir, 'input.txt')
            with open(input_file, 'w') as f:
                f.write(gene_list + '\n')
            ecode, out, err = self.run_enrichr_docker([input_file],
                                                      temp_dir=temp_dir)
            self.assertEqual(0, ecode)

            res = json.loads(out)
            self.assertEqual(8, len(res.keys()))
            self.assertEqual('receptor guanylyl cyclase signaling pathway (GO:0007168)', res['name'])
            self.assertEqual('GO_Biological_Process_2018', res['source'])
            self.assertEqual('', res['description'])
            self.assertEqual(10, res['term_size'])
            self.assertEqual(6, len(res['intersections']))

        finally:
            shutil.rmtree(temp_dir)


if __name__ == '__main__':
    sys.exit(unittest.main())


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'ndexbio/cdenrichrgenestoterm'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


cdenrichrgenestoterm/cdenrichrgenestoterm.py
#!/usr/bin/env python

import os
import sys
import argparse
import json
import pandas
from contextlib import redirect_stdout

with redirect_stdout(sys.stderr):
    import gseapy


ADJUSTED_PVALUE = 'Adjusted P-value'
PVALUE = 'P-value'

class Formatter(argparse.ArgumentDefaultsHelpFormatter, argparse.RawDescriptionHelpFormatter):
    pass


def _parse_arguments(desc, args):
    """
    Parses command line arguments
    :param desc:
    :param args:
    :return:
    """
    help_fm = Formatter
    parser = argparse.ArgumentParser(description=desc,
                                     formatter_class=help_fm)
    parser.add_argument('input',
                        help='comma delimited list of genes in file')
    parser.add_argument('--maxpval', type=float, default=0.05,
                        help='Max p value')
    parser.add_argument('--tmpdir', default='/tmp',
                        help='Temp directory to hold output from task')
    parser.add_argument('--genesets', default='GO_Biological_Process_2018,'
                                              'GO_Cellular_Component_2018,'
                                              'GO_Molecular_Function_2018',
                        help='Gene sets to enrich against. '
                             'Should be comma delimited')
    return parser.parse_args(args)


def read_inputfile(inputfile):
    """

    :param inputfile:
    :return:
    """
    with open(inputfile, 'r') as f:
        return f.read()


def load_data_frame_from_outputfiles(outdir=None):
    """
    Loads all files ending with `.txt` loading them
    into a single pandas data frame
    :param outdir:
    :return: combined data of .txt files into single pandas data frame
    :rtype: :py:class:`pandas.DataFrame`
    """
    d_frames = []
    for entry in os.listdir(outdir):
        if not entry.endswith('.txt'):
            continue
        full_path = os.path.join(outdir, entry)
        if not os.path.isfile(full_path):
            continue
        df = pandas.read_csv(full_path, delimiter='\t', header=0)
        d_frames.append(df)

    if len(d_frames) == 0:
        return pandas.DataFrame()

    mega_df = pandas.concat(d_frames)
    mega_df.reset_index(drop=True, inplace=True)
    return mega_df


def run_enrichr(inputfile, theargs,
                enrichr=gseapy,
                retry_count=2):
    """
    todo
    :param inputfile:
    :return:
    """
    genes = read_inputfile(inputfile)
    genes = genes.strip(',').strip('\n').upper().split(',')
    if genes is None or (len(genes) == 1 and len(genes[0].strip()) == 0):
        sys.stderr.write('No genes found in input')
        return None
    cur_try = 1
    with redirect_stdout(sys.stderr):
        while cur_try <= retry_count:
            try:
                enrichr.enrichr(gene_list=genes, gene_sets=theargs.genesets,
                                cutoff=theargs.maxpval,
                                no_plot=True, outdir=theargs.tmpdir)
                break
            except Exception as e:
                sys.stderr.write('Try # ' + str(cur_try) + ' caught exception: ' + str(e))
                cur_try += 1
                if cur_try > retry_count:
                    sys.stderr.write('Retries exceeded')
                    return None

    df_result = load_data_frame_from_outputfiles(outdir=theargs.tmpdir)
    if df_result.shape[0] == 0:
        sys.stderr.write('Empty data frame\n')
        return None
    """
    Example output:
    >>> df_result
          Gene_set                      Term  Overlap   P-value  Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  Combined Score Genes
    0  Jensen_DISEASES                 Keratosis     1/27  0.460766               1.0            0                     0    1.638807    1.269854e+00   TAT
    1  Jensen_DISEASES  Hepatocellular carcinoma     1/29  0.484897               1.0            0                     0    1.525786    1.104393e+00   TAT
    2  Jensen_DISEASES    Palmoplantar keratosis     1/31  0.507950               1.0            0                     0    1.427348    9.668460e-01   TAT
    3  Jensen_DISEASES                 Keratitis     1/39  0.590316               1.0            0                     0    1.134559    5.980234e-01   TAT
    4  Jensen_DISEASES   Intellectual disability    1/296  0.998903               1.0            0                     0    0.149486    1.641426e-04   TAT
    5  Jensen_DISEASES                 Carcinoma  1/11318  0.999993               1.0            0                     0    0.003910    2.732093e-08   TAT
    """
    # filter out any rows where min overlap is not met
    df_result.drop(df_result[df_result[ADJUSTED_PVALUE] > theargs.maxpval].index,
                   inplace=True)
    if df_result.shape[0] == 0:
        sys.stderr.write('Empty data frame after p value filter\n')
        return None
    df_result.sort_values([ADJUSTED_PVALUE, PVALUE],
                          ascending=True, inplace=True)

    df_result.reset_index(drop=True, inplace=True)
    theres = {'name': df_result['Term'][0],
              'source': df_result['Gene_set'][0],
              'sourceTermId': '',
              'p_value': df_result[ADJUSTED_PVALUE][0],
              'description': '',
              'term_size': int(df_result['Overlap'][0][df_result['Overlap'][0].index('/')+1:]),
              'intersections': df_result['Genes'][0].split(';')}
    theres['jaccard'] = round(len(theres['intersections'])/len(genes), 3)
    return theres


def main(args):
    """
    Main entry point for program

    :param args: command line arguments usually :py:const:`sys.argv`
    :return: 0 for success otherwise failure
    :rtype: int
    """
    desc = """
        Running Enrichr via gseapy 0.10.4 

        Takes file with comma delimited list of genes as input and
        outputs best matching term (as determined by Adjusted P value)
        if any in JSON format:
        
        {
         "name": "TERM",
         "source": "SOURCE OF TERM",
         "sourceTermId": "IS THE ID FOR THE ENRICHED TERM/FUNCTIONAL CATEGORY IN ITS NATIVE NAMESPACE"
         "p_value": Adjusted P-value from Enrichr,
         "description": "EMPTY STRING",
         "intersections": "List of Genes that intersect"
        }
        
    """

    theargs = _parse_arguments(desc, args[1:])

    try:
        inputfile = os.path.abspath(theargs.input)
        theres = run_enrichr(inputfile, theargs)
        sys.stderr.flush()
        if theres is None:
            sys.stderr.write('No terms found\n')
        else:
            json.dump(theres, sys.stdout)
        sys.stdout.flush()
        return 0
    except Exception as e:
        sys.stderr.write('Caught exception: ' + str(e))
        return 2
    finally:
        sys.stderr.flush()


if __name__ == '__main__':  # pragma: no cover
    sys.exit(main(sys.argv))
